### ISSUE TITLE
send_newsletter using the wrong ID for user IDs

### ISSUE THREAD

**Issue by vicotec** [Wunderbyte-GmbH/moodle-mod_newsletter#52]
**Plugin version:**
v2.3.5-NewYorkTimes (Build: 2023072700)

**Moodle Version:**
Moodle 4.1.5 (Build: 20230814)

**Location**
moodle-mod_newsletter/classes/task/send_newsletter.php line 101: ``$sub->userid = $userid;``

**the issue**
$userid in this case is the ID of the newsletter_subscription itself, not the user ID.

**solution**
Change the line into ``$sub->userid = $recipient->userid;``
Alternatively, rearrange the query in ``newsletter_get_all_valid_recipients``.

**Comment by gmanaud** [comment]
I had the same problem with Moodle 4.1.6 (Build: 20231009) and v2.3.5-NewYorkTimes (Build: 2023072700). Replacing $sub->userid = $userid; with $sub->userid = $recipient->userid; actually solved the problem.

**Comment by ChristopherFree** [comment]
I was experiencing a problem with this also. 
 
Versions
Moodle 4.2 (Build: 20230424) and Moodle 4.2.2 (Build: 20230814). 

The issue
I created a newsletter and subscribed two students. Only student-1 received the newsletter. I checked my cPanel email delivery and Moodle did not send an email to student-2.

When I added a student-3 and sent out a new edition of the newsletter, the student-1 and student-2 received the newsletter and student-3 did not receive the newsletter. I checked my cPanel email delivery and Moodle had sent emails to student-1 and student-2, but there was no record of an email being sent to student-3. 

In this limited test, it appears the bug was causing the newest subscriber to be left out. I did not test with longer lists of subscribers to see if there was another pattern to the dropped users. This is a new LMS install with a total of 9 user accounts. 

Solution: 
I used the code provided by gmanaud above and that solved the problem. All emails are sent and delivered as expected.  
$sub->userid = $userid; with $sub->userid = $recipient->userid;
